internal/uvm/network.go

```diff
@@ -1,31 +1,129 @@
 package uvm
 
 import (
-	"fmt"
+	"errors"
 	"path"
 
+	"github.com/sirupsen/logrus"
+
 	"github.com/Microsoft/hcsshim/hcn"
 	"github.com/Microsoft/hcsshim/internal/guestrequest"
 	"github.com/Microsoft/hcsshim/internal/guid"
 	"github.com/Microsoft/hcsshim/internal/hns"
+	"github.com/Microsoft/hcsshim/internal/logfields"
 	"github.com/Microsoft/hcsshim/internal/requesttype"
 	"github.com/Microsoft/hcsshim/internal/schema1"
 	"github.com/Microsoft/hcsshim/internal/schema2"
 	"github.com/Microsoft/hcsshim/osversion"
-	"github.com/sirupsen/logrus"
 )
 
-// AddNetNS adds network namespace inside the guest & adds endpoints to the guest on that namepace
-func (uvm *UtilityVM) AddNetNS(id string, endpoints []*hns.HNSEndpoint) (err error) {
+var (
+	// ErrNetNSAlreadyAttached is an error indicating the guest UVM already has
+	// an endpoint by this id.
+	ErrNetNSAlreadyAttached = errors.New("network namespace already added")
+	ErrNetNSNotFound        = errors.New("network namespace not found")
+)
+
+// AddNetNS adds network namespace inside the guest.
+//
+// If a namespace with `id` already exists returns `ErrNetNSAlreadyAttached`.
+func (uvm *UtilityVM) AddNetNS(id string) error {
+	logrus.WithFields(logrus.Fields{
+		logfields.UVMID: uvm.id,
+		"netns-id":      id,
+	}).Debug("uvm::AddNetNS")
+
+	uvm.m.Lock()
+	defer uvm.m.Unlock()
```

Comment:
Wondering if we should hold the lock throughout this relatively long-running operation, effectively blocking any other potential operation against the UVM.

Reply:
I'm honestly not sure why we hold the lock for any of these. Doesn't the platform lock this anyways? But if we remove the lock it will really complicate the model. For now can we leave the locks but revisit this for all scsi/pmem etc. I really think we should only hold the lock for `id` reservation and on failure it will release the held `id`.

Reply:
How much concurrency do we have with the new shim model? In the runhcs model, I don’t think we even relied on the locking...

Reply:
Also the old code held the lock this long as well. (Not that thats a good excuse) :)

Reply:
Honestly the new shim will only have exactly 1 uvm and the network is only provisioned at start time. So in this case absolutely zero concurrency in practice.

Reply:
I wouldn’t worry about the locking efficiency for now. Maybe file a cleanup task — perhaps the uvm contract should require single-threaded access in which case we could remove all the locks.

Reply:
Yep agreed. Needs revisiting